WcaOnRails/app/controllers/competitions_controller.rb

```diff
@@ -87,19 +87,41 @@ def index
     end
 
     if @display == 'admin'
-      @competitions = @competitions.includes(:delegates, :delegate_report)
+      @competitions = @competitions.includes(:delegate_report)
     end
 
     unless params[:region] == "all"
-      @competitions = @competitions.select { |competition| competition.belongs_to_region?(params[:region]) }
+      @competitions = @competitions.where("countryId = ? or continentId = ?", params[:region], params[:region])
     end
 
     if params[:search].present?
-      @competitions = @competitions.select { |competition| competition.contains?(params[:search]) }
+      params[:search].split.all? do |part|
+        @competitions = @competitions.where(
+          "Competitions.name like ? or
+          Competitions.cityName like ? or
+          users.name like ? or
+          Competitions.venue like ? or
+          Competitions.cellName like ? or
+          Competitions.countryId like ?",
+          "%#{part}%",
+          "%#{part}%",
+          "%#{part}%",
+          "%#{part}%",
+          "%#{part}%",
+          "%#{part}%",
+        )
+      end
     end
 
     unless params[:event_ids].empty?
-      @competitions = @competitions.select { |competition| competition.has_events_with_ids?(params[:event_ids]) }
+      params[:event_ids].each do |event_id|
+        ce_alias = "competition_events" + event_id
+        e_alias = "Events" + event_id
+        @competitions = @competitions.joins(
+          "join competition_events " + ce_alias + " ON " + ce_alias + ".competition_id = Competitions.id " \
+          "join Events " + e_alias + " ON " + e_alias + ".id = " + ce_alias + ".event_id",
+        ).where(e_alias + ".id = ?", event_id)
+      end
```

Comment:
My suspicion is the opposite.  Some of this code, especially the multiple joins, look like it asks a lot of MySQL. Meanwhile, `.select()`-ing an array in memory should be a sub ms operation. If it's normally 3000 competitions, I agree it's crazier than I thought.

But my guess comes from a deep ignorance of the schema and data. I'm just eye balling the code. Which is why measuring - like always when optimizing - is important.

Sadly, comparing DB query speed is devilishly hard. Since the DB caches the response data, the second query you try will **always** be the fastest. By a lot. There are a number of ways to work around that.

- Turn off or invalidate the DB cache somehow. I've tried following the confident web descriptions of how to do this several times. No luck.
- Have a set of thousands search queries, throw a random 100 of them at each implementation, and compare the averages. This worked well on Birdflu.
- Deploy both as an A/B test. Collect average times over a week. Feels "dirty", but should be fairly simple and safe.